pkg/yurthub/proxy/multiplexer/multiplexerproxy.go

```diff
@@ -0,0 +1,251 @@
+	"k8s.io/apiserver/pkg/endpoints/handlers"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	kstorage "k8s.io/apiserver/pkg/storage"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/yurthub/filter"
+	"github.com/openyurtio/openyurt/pkg/yurthub/multiplexer"
+)
+
+type multiplexerProxy struct {
+	cacheManager multiplexer.MultiplexerManager
+	filterMgr    filter.FilterManager
+	stop         <-chan struct{}
+}
+
+func NewDefaultShareProxy(filterMgr filter.FilterManager,
+	cacheManager multiplexer.MultiplexerManager,
+	multiplexerResources []schema.GroupVersionResource,
+	stop <-chan struct{}) (*multiplexerProxy, error) {
+
+	sp := &multiplexerProxy{
+		stop:         stop,
+		cacheManager: cacheManager,
+		filterMgr:    filterMgr,
+	}
+
+	for _, gvr := range multiplexerResources {
+		if _, _, err := sp.cacheManager.ResourceCache(&gvr); err != nil {
+			return sp, errors.Wrapf(err, "failed to init resource cache for %s", gvr.String())
+		}
+	}
+
+	return sp, nil
+}
+
+func (sp *multiplexerProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	reqInfo, _ := request.RequestInfoFrom(r.Context())
+	gvr := sp.getRequestGVR(reqInfo)
+
+	switch reqInfo.Verb {
+	case "list":
+		sp.multiplexerList(w, r, gvr)
+	case "watch":
+		sp.multiplexerWatch(w, r, gvr)
+	}
+}
+
+func (sp *multiplexerProxy) getRequestGVR(reqInfo *request.RequestInfo) *schema.GroupVersionResource {
+	return &schema.GroupVersionResource{
+		Group:    reqInfo.APIGroup,
+		Version:  reqInfo.APIVersion,
+		Resource: reqInfo.Resource,
+	}
+}
+
+func (sp *multiplexerProxy) getReqScope(gvr *schema.GroupVersionResource) (*handlers.RequestScope, error) {
+	namer, err := sp.getNamer(gvr)
+	if err != nil {
+		return nil, err
+	}
+
+	fqKindToRegister, err := sp.findKind(gvr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &handlers.RequestScope{
+		Serializer:      scheme.Codecs,
+		ParameterCodec:  scheme.ParameterCodec,
+		Convertor:       scheme.Scheme,
+		Defaulter:       scheme.Scheme,
+		Typer:           scheme.Scheme,
+		UnsafeConvertor: runtime.UnsafeObjectConvertor(scheme.Scheme),
+		Authorizer:      authorizerfactory.NewAlwaysAllowAuthorizer(),
+
+		EquivalentResourceMapper: runtime.NewEquivalentResourceRegistry(),
+
+		// TODO: Check for the interface on storage
+		TableConvertor: rest.NewDefaultTableConvertor(gvr.GroupResource()),
+
+		// TODO: This seems wrong for cross-group subresources. It makes an assumption that a subresource and its parent are in the same group version. Revisit this.
+		Resource: *gvr,
+		Kind:     fqKindToRegister,
+
+		HubGroupVersion: schema.GroupVersion{Group: fqKindToRegister.Group, Version: runtime.APIVersionInternal},
+
+		MetaGroupVersion: metav1.SchemeGroupVersion,
+
+		MaxRequestBodyBytes: int64(3 * 1024 * 1024),
+		Namer:               namer,
+	}, nil
+}
+
+func (sp *multiplexerProxy) getNamer(gvr *schema.GroupVersionResource) (handlers.ScopeNamer, error) {
```

Comment:
It seems that it is no need to new a func `getNamer`, you can only use `handlers.ContextBasedNaming{}` directly.

Reply:
done